### PR TITLE
Trombi

### PR DESCRIPTION
**Pull request by Piwakk**
Please merge #170 before

### CODE REVIEW

File: backend/authentication/views/profile.py [hunk 1]
```diff
@@ -1,26 +1,71 @@
...
 from authentication.permissions import ProfilePermission
 from authentication.serializers.user import (
     ReadOnlyUserSerializer,
     UpdateOnlyUserSerializer,
+    UserShortSerializer,
 )
+from tags.filters import HasHiddenTagFilter
+
+
+class ProfileFilter(FilterSet):
```
Comment (abocquet) [inline backend/authentication/views/profile.py:22]: Wouldn't `promotion__in` work?

File: backend/authentication/views/promotions.py [hunk 2]
```diff
@@ -6,11 +6,10 @@
...
-
-    promotions = [
-        x[0] % 100
-        for x in User.objects.distinct("year_of_entry").values_list("year_of_entry")
-    ]
-    return Response({"promotions": sorted(list(promotions), reverse=True)})
+    return Response(
+        {
+            "promotions": sorted(
+                set(User.objects.values_list("promotion", flat=True)), reverse=True
```
Comment (abocquet) [inline backend/authentication/views/promotions.py:12]: Use `distinct` instead https://docs.djangoproject.com/en/2.2/ref/models/querysets/#distinct

File: frontend/src/components/user/trombi/Trombi.tsx [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        }
+    }, [promotions]);
+
+    // The search key sent to `api.users.list`. It is debounced thanks to the `searchInputValue` hooks.
+    const [searchKey, setSearchKey] = useState<string>("");
+    const [searchInputValue, setSearchInputValue] = useState<string>("");
+
+    useEffect(() => {
+        const timer = setTimeout(() => {
+            setSearchKey(searchInputValue);
```
Comment (abocquet) [inline frontend/src/components/user/trombi/Trombi.tsx:42]: Use `react-debounce-input` instead (https://www.npmjs.com/package/react-debounce-input). It's already imported in another branch so don't worry about installing it here

File: frontend/src/components/utils/avatar/UserAvatarCard.tsx [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+}: {
+    children?: any;
+    className?: string;
+    avatarClassName?: string;
+} & UserAvatarProps) => (
+    <Card className={`justify-content-center ${className}`}>
+        <UserAvatar
+            userId={userId}
+            className={avatarClassName}
+            linkClassName={"mx-auto " + (linkClassName ? linkClassName : "")}
```
Comment (abocquet) [inline frontend/src/components/utils/avatar/UserAvatarCard.tsx:31]:
```suggestion
            linkClassName={"mx-auto " + (linkClassName || "")}
```

File: frontend/src/components/utils/avatar/UserAvatar.tsx [hunk 5]
```diff
@@ -40,7 +45,7 @@ export const UserAvatar = ({
         />
     );
     return link ? (
-        <Link to={`/profils/${userId}`} className="user-avatar-link">
+        <Link to={`/profils/${userId}`} className={`user-avatar-link ${linkClassName ? linkClassName : ""}`}>
```
Comment (abocquet) [inline frontend/src/components/utils/avatar/UserAvatar.tsx:48]:
```suggestion
        <Link to={`/profils/${userId}`} className={`user-avatar-link ${linkClassName || ""}`}>
```

File: frontend/src/services/apiService.ts [hunk 6]
```diff
@@ -68,6 +69,7 @@ export const api = {
     polls: polls,
     tags: tags,
     namespaces: namespaces,
+    users:users,
```
Comment (abocquet) [inline frontend/src/services/apiService.ts:72]:
```suggestion
    users: users,
```
I'm surprised Prettier lets this pass

File: backend/authentication/views/promotions.py (hunk 2 above)
Comment (Piwakk) [reply backend/authentication/views/promotions.py:12]: It gave me a weird SQL error… I will try again

File: frontend/src/services/apiService.ts (hunk 6 above)
Comment (Piwakk) [reply frontend/src/services/apiService.ts:72]: I think the CI is broken (again)

File: backend/authentication/views/profile.py [hunk 7]
```diff
@@ -1,26 +1,66 @@
...
+class CharInFilter(BaseInFilter, CharFilter):
+    pass
+
+
+class ProfileFilter(FilterSet):
+    """
+    This class is needed because of the behaviour of the `promotion` field: the commas (`,`) act as OR.
+    """
+
+    promotion__in = CharInFilter(field_name="promotion", lookup_expr="in")
```
Comment (abocquet) [inline backend/authentication/views/profile.py:31]:
```suggestion
```

File: backend/authentication/views/profile.py [hunk 8]
```diff
@@ -1,26 +1,66 @@
...
+
+    class Meta:
+        model = User
+        fields = ("promotion__in",)
```
Comment (abocquet) [inline backend/authentication/views/profile.py:35]:
```suggestion
        fields = {"promotion": ["exact", "in"]}
```

File: backend/authentication/views/promotions.py [hunk 9]
```diff
@@ -6,11 +6,15 @@
...
 def list_promotions(request):
-    # Technically speaking, we are not sure at 100 % to have every promotion.
-    # In practice, it will always be OK.
-
-    promotions = [
-        x[0] % 100
-        for x in User.objects.distinct("year_of_entry").values_list("year_of_entry")
-    ]
-    return Response({"promotions": sorted(list(promotions), reverse=True)})
+    # FWIW, `distinct("promotion")` only works on PostgreSQL.
```
Comment (abocquet) [inline backend/authentication/views/promotions.py:9]: PostgreSQL FTW

File: frontend/src/components/user/trombi/Trombi.tsx [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+    useEffect(() => {
+        if (promotions) {
+            setPromotionsOptions(
+                promotions.promotions.map((promotion) => {
+                    return { value: promotion, label: promotion };
+                })
+            );
+        }
+    }, [promotions]);
```
Comment (abocquet) [inline frontend/src/components/user/trombi/Trombi.tsx:34]:
```suggestion
   
```

File: frontend/src/components/user/trombi/Trombi.tsx [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+                            <span className="input-icon-addon">
+                                <i className="fe fe-search"/>
+                            </span>
+                        </div>
+                    </InputGroup>
+                </Col>
+                <Col md={{ span: 3, offset: 6 }}>
+                    <Select
+                        value={promotionsFilter}
+                        options={promotionsOptions}
```
Comment (abocquet) [inline frontend/src/components/user/trombi/Trombi.tsx:70]:
```suggestion
                        options={promotions.promotions.map((promotion) =>({ value: promotion, label: promotion })}
```

File: frontend/src/components/user/trombi/Trombi.tsx [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+            setPromotionsOptions(
+                promotions.promotions.map((promotion) => {
+                    return { value: promotion, label: promotion };
+                })
+            );
+        }
+    }, [promotions]);
+
+    // The search key sent to `api.users.list`. It is debounced thanks to the `searchInputValue` hooks.
+    const [searchKey, setSearchKey] = useState<string>("");
```
Comment (abocquet) [inline frontend/src/components/user/trombi/Trombi.tsx:37]: TS compiler is able to infer the type from default value
```suggestion
    const [searchKey, setSearchKey] = useState("");
```

File: frontend/src/components/user/trombi/Trombi.tsx [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+                <Col md={{ span: 3, offset: 6 }}>
+                    <Select
+                        value={promotionsFilter}
+                        options={promotionsOptions}
+                        closeMenuOnSelect={false}
+                        isMulti
+                        placeholder="Filtrer par promotion…"
+                        onChange={(newPromotions) =>
+                            setPromotionsFilter(newPromotions)
+                        }
```
Comment (abocquet) [inline frontend/src/components/user/trombi/Trombi.tsx:76]:
```suggestion
                        onChange={setPromotionsFilter}
```

File: frontend/src/components/user/trombi/Trombi.tsx [hunk 14]
```diff
@@ -59,21 +44,28 @@ export const Trombi = () => {
...
                         value={promotionsFilter}
-                        options={promotionsOptions}
+                        options={
+                            promotions
+                                ? promotions.promotions.map((promotion) => {
+                                      return {
+                                          value: promotion,
+                                          label: promotion,
+                                      };
+                                  })
```
Comment (abocquet) [inline frontend/src/components/user/trombi/Trombi.tsx:62]:
```suggestion
                                ? promotions.promotions.map((promotion) => ({
                                          value: promotion,
                                          label: promotion
                                  }))
```

**Pull request merged**